### PR TITLE
Reuse UnboundedReaders in the InProcessRunner

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by tgroh** [comment]
R: @dhalperi

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/UnboundedReadEvaluatorFactory.java [hunk 1]
```diff
@@ -146,35 +170,87 @@ public void processElement(WindowedValue<Object> element) {}
...
-                    reader.getCurrent(), reader.getCurrentTimestamp()));
+            output.add(WindowedValue.timestampedValueInGlobalWindow(currentReader.getCurrent(),
+                currentReader.getCurrentTimestamp()));
             numElements++;
-          } while (numElements < ARBITRARY_MAX_ELEMENTS && reader.advance());
+          } while (numElements < ARBITRARY_MAX_ELEMENTS && currentReader.advance());
+          watermark = currentReader.getWatermark();
+          // Only take a checkpoint if we did any work
+          finishRead();
         }
```
Comment (dhalperi) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/UnboundedReadEvaluatorFactory.java]: move finalize down here? We don't want users somehow relying on finalize happening immediately.

Comment (tgroh) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/UnboundedReadEvaluatorFactory.java]: Two different locations;

When taking a new checkpoint, or when resuming from a checkpoint.

Cleared when resuming so it isn't finalized twice

Comment (dhalperi) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/UnboundedReadEvaluatorFactory.java]: it looks like you didn't push?

**Comment by dhalperi** [comment]
LGTM

**Pull request closed without merging**